bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/FindReplaceLogic.java

```diff
@@ -0,0 +1,687 @@
+		int index = 0;
+		if (isActive(SearchOptions.FORWARD)) {
+			index = findAndSelect(startPosition, findString);
+		} else {
+			index = startPosition == 0 ? -1
+				: findAndSelect(startPosition - 1, findString);
+		}
+
+		if (index == -1) {
+
+			if (isActive(SearchOptions.WRAP)) {
+				statusLineMessage(FindReplaceMessages.FindReplace_Status_wrapped_label);
+				status = new FindStatus(FindStatus.StatusCode.WRAPPED, false);
+				index = findAndSelect(-1, findString);
+			} else {
+				status = new FindStatus(FindStatus.StatusCode.NO_MATCH, false);
+			}
+		}
+		return index;
+	}
+
+	@Override
+	public int findAndSelect(int offset, String findString) {
+		if (target instanceof IFindReplaceTargetExtension3)
+			return ((IFindReplaceTargetExtension3) target).findAndSelect(offset, findString,
+					isActive(SearchOptions.FORWARD),
+					isActive(SearchOptions.CASE_SENSITIVE), isWholeWordSearchAvailableAndActive(), isActive(SearchOptions.REGEX));
+		return target.findAndSelect(offset, findString, isActive(SearchOptions.FORWARD),
+				isActive(SearchOptions.CASE_SENSITIVE),
+				isWholeWordSearchAvailableAndActive());
+	}
+
+	/**
+	 * Replaces the selection with <code>replaceString</code>. If
+	 * <code>regExReplace</code> is <code>true</code>, <code>replaceString</code> is
+	 * a regex replace pattern which will get expanded if the underlying target
+	 * supports it. Returns the region of the inserted text; note that the returned
+	 * selection covers the expanded pattern in case of regex replace.
+	 *
+	 * @param replaceString the replace string (or a regex pattern)
+	 * @param regExReplace  <code>true</code> if <code>replaceString</code> is a
+	 *                      pattern
+	 * @return the selection after replacing, i.e. the inserted text
+	 */
+	private Point replaceSelection(String replaceString, boolean regExReplace) {
+		if (target instanceof IFindReplaceTargetExtension3)
+			((IFindReplaceTargetExtension3) target).replaceSelection(replaceString, regExReplace);
+		else
+			target.replaceSelection(replaceString);
+
+		return target.getSelection();
+	}
+
+	/**
+	 * Returns whether the specified search string can be found using the given
+	 * options.
+	 *
+	 * @param findString             the string to search for
+	 * @param forwardSearch          the direction of the search
+	 * @return <code>true</code> if the search string can be found using the given
+	 *         options
+	 *
+	 */
+	private boolean findNext(String findString, boolean forwardSearch) {
+
+		if (target == null) {
+			return false;
+		}
+
+		Point r = null;
+		if (isActive(SearchOptions.INCREMENTAL)) {
+			r = incrementalBaseLocation;
+		} else {
+			r = target.getSelection();
+		}
+
+		int findReplacePosition = r.x;
+		if (forwardSearch // && !nextReplaceOperationNeedsFindOperationFirst
+		) { // || !forwardSearch && nextReplaceOperationNeedsFindOperationFirst) {
+			findReplacePosition += r.y;
+		}
+
+		// nextReplaceOperationNeedsFindOperationFirst = false;
```

Comment:
Please remove commented out code if it is not necessary.

Reply:
https://github.com/eclipse-platform/eclipse.platform.ui/commit/c5665ccee9436e10b611c828ebcf8f501d0b189f